File: program-libs/interface/src/shape.rs
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct Shape {
    n_inputs: usize,
    n_outputs: usize,
}

impl Shape {
    pub const IN1_OUT1: Self = Self {
        n_inputs: 1,
        n_outputs: 1,
    };
    pub const IN1_OUT2: Self = Self {
        n_inputs: 1,
        n_outputs: 2,
    };
    pub const IN2_OUT2: Self = Self {
        n_inputs: 2,
        n_outputs: 2,
    };
    pub const IN2_OUT3: Self = Self {
        n_inputs: 2,
        n_outputs: 3,
    };
    pub const IN3_OUT3: Self = Self {
        n_inputs: 3,
        n_outputs: 3,
    };
    pub const IN4_OUT3: Self = Self {
        n_inputs: 4,
        n_outputs: 3,
    };
    pub const IN4_OUT4: Self = Self {
        n_inputs: 4,
        n_outputs: 4,
    };
    pub const IN5_OUT3: Self = Self {
        n_inputs: 5,
        n_outputs: 3,
    };
    pub const IN5_OUT4: Self = Self {
        n_inputs: 5,
        n_outputs: 4,
    };
    pub const IN1_OUT8: Self = Self {
        n_inputs: 1,
        n_outputs: 8,
    };
    pub const IN36_OUT2: Self = Self {
        n_inputs: 36,
        n_outputs: 2,
    };

    pub const fn new(n_inputs: usize, n_outputs: usize) -> Self {
        Self {
            n_inputs,
            n_outputs,
        }
    }

    pub const fn n_inputs(&self) -> usize {
        self.n_inputs
    }

    pub const fn n_outputs(&self) -> usize {
        self.n_outputs
    }

    /// Slots in the public signer vector on the signature-requiring rails:
    /// the payer followed by [`owner_signer_slots`] owner signers. The ring
    /// authority rail publishes a single slot instead.
    pub const fn signer_width(self) -> usize {
        owner_signer_slots(self.n_inputs) + 1
    }
}

/// Distinct addresses one transaction can carry, `solana_message::v1::MAX_ADDRESSES`.
pub const MAX_TRANSACTION_ADDRESSES: usize = 64;

/// Addresses every `transact` needs besides its nullifier PDAs and owner
/// signers: payer, input tree (the output tree may coincide with it), the
/// shielded-pool program and the system program.
pub const FIXED_TRANSACT_ADDRESSES: usize = 4;

/// Owner signer slots the public signer vector reserves for `n_inputs` inputs.
///
/// One per input, capped by the addresses a transaction has left after the
/// fixed accounts and one nullifier PDA per input. The circuit width has to be
/// the bound the program cannot rule out; PDA owners sign through CPI, so the
/// transaction signature cap does not lower it, and settlement accounts only
/// reduce the reachable count further.
pub const fn owner_signer_slots(n_inputs: usize) -> usize {
    let remaining = MAX_TRANSACTION_ADDRESSES.saturating_sub(FIXED_TRANSACT_ADDRESSES + n_inputs);
    if n_inputs < remaining {
        n_inputs
    } else {
        remaining
    }
}

/// Widest public signer vector over [`SPP_SUPPORTED_SHAPES`]; the program sizes
/// its fixed signer buffers from it.
pub const MAX_SIGNERS: usize = 25;

/// The buffer bounds hold for every supported shape, or the build fails.
const _: () = {
    let mut shapes: &[Shape] = &SPP_SUPPORTED_SHAPES;
    while let Some((shape, rest)) = shapes.split_first() {
        assert!(shape.signer_width() <= MAX_SIGNERS);
        assert!(shape.n_inputs <= crate::MAX_TRANSACT_INPUTS);
        assert!(shape.n_outputs <= crate::MAX_OUTPUTS);
        shapes = rest;
    }
};

/// Shapes the SPP prover has keys for. Slot-signed transactions declare their
/// exact shape (they do not pad), so they validate against this full set rather
/// than the fixed padded-transfer shape ([`Shape::IN2_OUT3`]).
pub const SPP_SUPPORTED_SHAPES: [Shape; 11] = [
    Shape::IN1_OUT1,
    Shape::IN1_OUT2,
    Shape::IN2_OUT2,
    Shape::IN2_OUT3,
    Shape::IN3_OUT3,
    Shape::IN4_OUT3,
    Shape::IN4_OUT4,
    Shape::IN5_OUT3,
    Shape::IN5_OUT4,
    Shape::IN1_OUT8,
    Shape::IN36_OUT2,
];

File: sdk-libs/ts/src/interface/shape.ts
import { InterfaceError } from "./errors.js";

export type Shape = Readonly<{
  inputs: number;
  outputs: number;
}>;

function shape(inputs: number, outputs: number): Shape {
  return Object.freeze({ inputs, outputs });
}

export const SPP_SUPPORTED_SHAPES: readonly Shape[] = Object.freeze([
  shape(1, 1),
  shape(1, 2),
  shape(2, 2),
  shape(2, 3),
  shape(3, 3),
  shape(4, 3),
  shape(4, 4),
  shape(5, 3),
  shape(5, 4),
  shape(1, 8),
]);

function count(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new InterfaceError("INTERFACE_INVALID_SHAPE", { [name]: value });
  }
  return value;
}

/** Distinct addresses one transaction can carry, `solana_message::v1::MAX_ADDRESSES`. */
export const MAX_TRANSACTION_ADDRESSES = 64;

/**
 * Addresses every transact needs besides its nullifier accounts and owner
 * signers: payer, input tree (the output tree may coincide with it), the
 * shielded-pool program and the system program.
 */
export const FIXED_TRANSACT_ADDRESSES = 4;

/**
 * Owner signer slots the public signer vector reserves for `inputs` inputs: one
 * per input, capped by the addresses a transaction has left after the fixed
 * accounts and one nullifier account per input.
 */
export function ownerSignerSlots(inputs: number): number {
  count(inputs, "inputs");
  const remaining = Math.max(0, MAX_TRANSACTION_ADDRESSES - FIXED_TRANSACT_ADDRESSES - inputs);
  return Math.min(inputs, remaining);
}

/** Slots in the public signer vector: the payer followed by the owner signer slots. */
export function signerWidth(shape: Shape): number {
  return ownerSignerSlots(shape.inputs) + 1;
}

export function selectSppShape(inputs: number, outputs: number): Shape {
  count(inputs, "inputs");
  count(outputs, "outputs");
  const selected = SPP_SUPPORTED_SHAPES.find(
    (candidate) => inputs <= candidate.inputs && outputs <= candidate.outputs,
  );
  if (selected === undefined) {
    throw new InterfaceError("INTERFACE_INVALID_SHAPE", { inputs, outputs });
  }
  return selected;
}

export function validateSppShape(inputs: number, outputs: number, declared: Shape): Shape {
  count(inputs, "inputs");
  count(outputs, "outputs");
  count(declared.inputs, "declaredInputs");
  count(declared.outputs, "declaredOutputs");
  const canonical = SPP_SUPPORTED_SHAPES.find(
    (candidate) => candidate.inputs === declared.inputs && candidate.outputs === declared.outputs,
  );
  if (canonical === undefined || inputs > declared.inputs || outputs > declared.outputs) {
    throw new InterfaceError("INTERFACE_INVALID_SHAPE", {
      inputs,
      outputs,
      declaredInputs: declared.inputs,
      declaredOutputs: declared.outputs,
    });
  }
  return canonical;
}
